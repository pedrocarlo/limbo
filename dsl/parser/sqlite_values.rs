use chumsky::prelude::*;
use rusqlite::types::Value;

pub(super) fn integer<'src>() -> impl Parser<'src, &'src str, i64, extra::Err<Rich<'src, char>>> {
    let number = just('-')
        .or_not()
        .then(text::int(10))
        .to_slice()
        .map(|s: &str| s.parse().unwrap())
        .boxed();
    number
}

pub(super) fn real<'src>() -> impl Parser<'src, &'src str, f64, extra::Err<Rich<'src, char>>> {
    let digits = text::digits(10).to_slice();

    let frac = just('.').then(digits);

    let exp = just('e')
        .or(just('E'))
        .then(one_of("+-").or_not())
        .then(digits);

    let number = just('-')
        .or_not()
        .then(text::int(10))
        .then(frac)
        .then(exp.or_not())
        .to_slice()
        .map(|s: &str| s.parse().unwrap())
        .boxed();
    choice((
        number,
        just("Inf").to(core::f64::INFINITY),
        just("-Inf").to(core::f64::NEG_INFINITY),
    ))
}

pub(super) fn value_parser<'src>(
) -> impl Parser<'src, &'src str, Value, extra::Err<Rich<'src, char>>> {
    let boolean = choice((
        just("true").to(Value::Integer(1)),
        just("false").to(Value::Integer(0)),
    ))
    .boxed();

    choice((
        boolean,
        real().map(|f| Value::Real(f)),
        integer().map(|i| Value::Integer(i)),
        just("Null").to(Value::Null),
    ))
}

pub(super) fn sqlite_values_parser<'src>(
) -> impl Parser<'src, &'src str, Vec<Vec<Value>>, extra::Err<Rich<'src, char>>> {
    let value_parser = value_parser();

    let value_list = value_parser
        .separated_by(just(',').padded())
        .allow_trailing()
        .collect::<Vec<_>>()
        .delimited_by(just('[').padded(), just(']').padded())
        .boxed();

    let value_list_2d = value_list
        .separated_by(just(',').padded())
        .allow_trailing()
        .collect::<Vec<_>>()
        .delimited_by(just('[').padded(), just(']').padded())
        .boxed();

    choice((just("None").to(vec![vec![]]), value_list_2d))
}

#[cfg(test)]
mod tests {
    use chumsky::Parser;

    use crate::{
        assert_debug_snapshot_with_input,
        sqlite_values::{sqlite_values_parser, value_parser},
    };

    #[test]
    fn test_none_values() {
        let parser = sqlite_values_parser();
        let input = "None";
        let val = parser.parse(input).unwrap();
        assert_debug_snapshot_with_input!(input, val);
    }

    #[test]
    fn test_null_value() {
        let parser = value_parser();
        let input = "Null";
        let val = parser.parse(input).unwrap();
        assert_debug_snapshot_with_input!(input, val);

        let input = "null";
        let _ = parser.parse(input).into_result().unwrap_err();
    }

    #[test]
    fn test_boolean_value() {
        let parser = value_parser();
        let input = "true";
        let val = parser.parse(input).unwrap();
        assert_debug_snapshot_with_input!(input, val);

        let input = "false";
        let val = parser.parse(input).unwrap();
        assert_debug_snapshot_with_input!(input, val);
    }

    #[test]
    fn test_integer_value() {
        let parser = value_parser();
        let input = "1";
        let val = parser.parse(input).unwrap();
        assert_debug_snapshot_with_input!(input, val);

        let input = "0";
        let val = parser.parse(input).unwrap();
        assert_debug_snapshot_with_input!(input, val);

        let input = "-1";
        let val = parser.parse(input).unwrap();
        assert_debug_snapshot_with_input!(input, val);

        let input = "-1000021900";
        let val = parser.parse(input).unwrap();
        assert_debug_snapshot_with_input!(input, val);

        let input = "112343543009010293";
        let val = parser.parse(input).unwrap();
        assert_debug_snapshot_with_input!(input, val);
    }

    #[test]
    fn test_real_value() {
        let parser = value_parser();
        let input = "1.0";
        let val = parser.parse(input).unwrap();
        assert_debug_snapshot_with_input!(input, val);

        let input = "0.0";
        let val = parser.parse(input).unwrap();
        assert_debug_snapshot_with_input!(input, val);

        let input = "-1.0";
        let val = parser.parse(input).unwrap();
        assert_debug_snapshot_with_input!(input, val);

        let input = "-1000021900.0";
        let val = parser.parse(input).unwrap();
        assert_debug_snapshot_with_input!(input, val);

        let input = "112343543009010293.0";
        let val = parser.parse(input).unwrap();
        assert_debug_snapshot_with_input!(input, val);

        let input = "123.3489534";
        let val = parser.parse(input).unwrap();
        assert_debug_snapshot_with_input!(input, val);

        let input = "-123.3489534";
        let val = parser.parse(input).unwrap();
        assert_debug_snapshot_with_input!(input, val);

        let input = "-123.3489534E5";
        let val = parser.parse(input).unwrap();
        assert_debug_snapshot_with_input!(input, val);

        let input = "-123.3489534E-5";
        let val = parser.parse(input).unwrap();
        assert_debug_snapshot_with_input!(input, val);

        let input = "Inf";
        let val = parser.parse(input).unwrap();
        assert_debug_snapshot_with_input!(input, val);

        let input = "-Inf";
        let val = parser.parse(input).unwrap();
        assert_debug_snapshot_with_input!(input, val);
    }
}
