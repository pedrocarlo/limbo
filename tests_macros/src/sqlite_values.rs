use quote::quote;
use quote::ToTokens;
use quote::TokenStreamExt as _;
use rusqlite::types::Value;
use syn::token::Bracket;
use syn::{punctuated::Punctuated, Ident, Lit, Token};

pub fn sqlite_values_impl(input: ValueList2D) -> syn::Result<proc_macro2::TokenStream> {
    Ok(input.into_token_stream())
}

#[derive(Debug)]
struct NewValue(Value);

impl syn::parse::Parse for NewValue {
    fn parse(input: syn::parse::ParseStream) -> syn::Result<Self> {
        let value = if input.peek(Ident) {
            let ident: Ident = input.parse()?;
            let name = ident.to_string();
            match name.as_str() {
                null if null.eq_ignore_ascii_case("null") => Value::Null,
                _ => {
                    return Err(syn::Error::new(ident.span(), "expected NULL identifier"));
                }
            }
        } else {
            let curr: Lit = input.parse()?;
            let value = match curr {
                Lit::Bool(bool) => bool.value().into(),
                Lit::Str(s) => s.value().into(),
                Lit::Int(i) => i.base10_parse::<i64>()?.into(),
                Lit::Float(f) => f.base10_parse::<f64>()?.into(),
                Lit::ByteStr(b) => {
                    let hex = hex::decode(b.value())
                        .map_err(|e| syn::Error::new(b.span(), e.to_string()))?;
                    hex.into()
                }
                lit => {
                    return Err(syn::Error::new(lit.span(), "unexpected literal type"));
                }
            };
            value
        };
        Ok(Self(value))
    }
}

#[derive(Debug)]
struct ValueList {
    values: Vec<NewValue>,
}

impl syn::parse::Parse for ValueList {
    fn parse(input: syn::parse::ParseStream) -> syn::Result<Self> {
        let mut values = Vec::new();

        if input.peek(Bracket) {
            let content;
            syn::bracketed!(content in input);
            if content.is_empty() {
                return Ok(Self { values });
            }

            while !content.is_empty() {
                let value: NewValue = content.parse()?;
                values.push(value);
                if content.peek(Token![,]) {
                    content.parse::<Token![,]>()?;
                }
            }
        } else {
            let value: NewValue = input.parse()?;
            values.push(value);
        }

        Ok(Self { values })
    }
}

impl quote::ToTokens for ValueList {
    fn to_tokens(&self, tokens: &mut proc_macro2::TokenStream) {
        let iter = self.values.iter().map(|v| match &v.0 {
            Value::Null => quote! {
                ::rusqlite::types::Value::Null
            },
            Value::Integer(i) => quote! {
                ::rusqlite::types::Value::Integer(#i)
            },
            Value::Real(f) => quote! {
                ::rusqlite::types::Value::Real(#f)
            },
            Value::Text(t) => quote! {
                ::rusqlite::types::Value::Text(#t)
            },
            Value::Blob(b) => {
                quote! {
                    ::rusqlite::types::Value::Blob(vec![#(#b),*])
                }
            }
        });
        tokens.append_all(quote! { vec![#(#iter),*] });
    }
}

pub(crate) struct ValueList2D {
    inner: Punctuated<ValueList, Token![,]>,
}

impl syn::parse::Parse for ValueList2D {
    fn parse(input: syn::parse::ParseStream) -> syn::Result<Self> {
        if input.peek(Ident) {
            let ident: Ident = input.parse()?;
            if ident.to_string() != "None" {
                return Err(syn::Error::new(ident.span(), "expected None identifier for empty sqlite values"));
            }
            let punctuated = Punctuated::new();
            return Ok(Self { inner: punctuated })
        }

        if input.peek(Bracket) {
            let fork = input.fork();
            let first_content;
            let _first_bracket = syn::bracketed!(first_content in fork);
            if first_content.peek(Bracket) {
                let content;
                syn::bracketed!(content in input);
                let inner = Punctuated::parse_terminated(&content)?;
                return Ok(Self { inner });
            }
        }
        let inner = Punctuated::parse_terminated(input)?;
        Ok(Self { inner })
    }
}

impl quote::ToTokens for ValueList2D {
    fn to_tokens(&self, tokens: &mut proc_macro2::TokenStream) {
        let iter = self.inner.iter().map(|v| v.into_token_stream());
        tokens.append_all(quote! { vec![#(#iter),*] })
    }
}
