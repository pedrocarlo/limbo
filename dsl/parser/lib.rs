mod sqlite_values;

use std::borrow::Cow;

use chumsky::prelude::*;
use rusqlite::types::Value;
use sqlite_values::sqlite_values_parser;

#[derive(Debug, PartialEq)]
pub struct Test<'a> {
    /// List of databases to test on
    kind: TestKind<'a>,
    /// Test mode
    mode: TestMode,
    /// Test Name
    // Idea of a Cow<'a, str> is that you can manipulate the ident with format! and still use this struct
    // e.g converting the ident to snake_case
    ident: Cow<'a, str>,
    /// Sql Statement(s) to run
    statement: Statement<'a>,
    /// Expected Output Value from test
    values: Vec<Vec<Value>>,
}

#[derive(Debug, PartialEq)]
pub enum TestKind<'a> {
    /// Default Databases
    Default,
    /// Specific Databases
    Databases(Vec<&'a str>),
    /// In-memory Database
    Memory,
    /// In-memory Regex test
    Regex,
}

#[derive(Debug, PartialEq, Clone, Copy, Default)]
pub enum TestMode {
    /// Follow normal execution
    #[default]
    Normal,
    /// Expect an Error to occur
    Error,
}

/// Sql Statement(s)
#[derive(Debug, PartialEq, Eq)]
pub enum Statement<'a> {
    Single(&'a str),
    Many(Vec<&'a str>),
}

pub fn test_contents_without_values<'src>(
) -> impl Parser<'src, &'src str, (&'src str, Statement<'src>), extra::Err<Rich<'src, char>>> {
    let ident = text::unicode::ident().padded();

    let sql_query = text::unicode::ident();

    let statement = choice((
        sql_query.map(|s: &'src str| Statement::Single(s)),
        sql_query
            .separated_by(just(',').padded())
            .allow_trailing()
            .collect::<Vec<_>>()
            .delimited_by(just('[').padded(), just(']').padded())
            .map(|s: Vec<&str>| Statement::Many(s)),
    ))
    .boxed();

    let contents = ident
        .then_ignore(just(',').padded())
        .then(statement)
        .boxed();
    contents
}

pub fn test_parser<'src>() -> impl Parser<'src, &'src str, Test<'src>, extra::Err<Rich<'src, char>>>
{
    let test_keyword = text::keyword("test");
    let test_error_keyword = text::keyword("test_error");

    let contents_with_value = test_contents_without_values()
        .then(
            just(',')
                .padded()
                .ignore_then(sqlite_values_parser())
                .or_not(),
        )
        .map(|((ident, statement), values)| Test {
            kind: TestKind::Default,
            mode: TestMode::Normal,
            ident: ident.into(),
            statement,
            values: values.unwrap_or(vec![vec![]]),
        })
        .delimited_by(just('(').padded(), just(')').padded())
        .boxed();

    let contents_no_value = test_contents_without_values()
        .map(|(ident, statement)| Test {
            kind: TestKind::Default,
            mode: TestMode::Error,
            ident: ident.into(),
            statement,
            values: vec![vec![]],
        })
        .delimited_by(just('(').padded(), just(')').padded())
        .boxed();

    choice((
        test_keyword.ignore_then(contents_with_value),
        test_error_keyword.ignore_then(contents_no_value),
    ))
}

pub fn test_parser_many<'src>(
) -> impl Parser<'src, &'src str, Vec<Test<'src>>, extra::Err<Rich<'src, char>>> {
    test_parser()
        .padded()
        .repeated()
        .collect::<Vec<_>>()
        .boxed()
}

#[cfg(test)]
mod tests {
    use chumsky::Parser;

    use crate::{test_parser, test_parser_many};

    #[macro_export]
    macro_rules! assert_debug_snapshot_with_input {
        ($input:expr, $result:expr) => {
            ::insta::with_settings!({
                description => &format!("input: {}", $input)
            }, {
                ::insta::assert_debug_snapshot!($result);
            });
        };
    }

    #[test]
    fn test_single_statement() {
        let parser = test_parser();
        let input = "test(test_single, SELECT)";
        let res = parser.parse(input).unwrap();
        assert_debug_snapshot_with_input!(input, res);
    }

    #[test]
    fn test_many_statements_1() {
        let parser = test_parser();
        let input = "test(test_many, [SELECT,])";
        let res = parser.parse(input).unwrap();
        assert_debug_snapshot_with_input!(input, res);
    }

    #[test]
    fn test_many_statements_2() {
        let parser = test_parser();
        let input = "test(test_many, [SELECT, INSERT, DELETE])";
        let res = parser.parse(input).unwrap();
        assert_debug_snapshot_with_input!(input, res);
    }

    #[test]
    fn test_many_tests() {
        let parser = test_parser_many();
        let input = r#"
            test(test_many, [SELECT, INSERT, DELETE])
            test(test_many_2, [SELECT,])
            test(test_single, SELECT)
        "#;
        let res = parser.parse(input).unwrap();
        assert_debug_snapshot_with_input!(input, res);
    }

    #[test]
    fn test_single_statement_with_value() {
        let parser = test_parser();
        let input = "test(test_single, SELECT, [Null, 1])";
        let res = parser.parse(input).unwrap();
        assert_debug_snapshot_with_input!(input, res);
    }

    #[test]
    fn test_many_tests_with_value() {
        let parser = test_parser_many();
        let input = r#"
            test(test_many, [SELECT, INSERT, DELETE], [Null, 1])
            test(test_many_2, [SELECT,], [[Null, 1], ["hi"]])
            test(test_single, SELECT, 1.234)
        "#;
        let res = parser.parse(input).unwrap();
        assert_debug_snapshot_with_input!(input, res);
    }

    #[test]
    fn test_error() {
        let parser = test_parser();
        let input = r#"test_error(test_error, SELECT)"#;
        let res = parser.parse(input).unwrap();
        assert_debug_snapshot_with_input!(input, res);

        let input = r#"test_error(test_error, [SELECT, INSERT, DELETE])"#;
        let res = parser.parse(input).unwrap();
        assert_debug_snapshot_with_input!(input, res);
    }
}
