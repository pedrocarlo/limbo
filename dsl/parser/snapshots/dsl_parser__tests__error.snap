---
source: dsl/parser/lib.rs
description: "input: test_error(test_error, SELECT)"
expression: res
---
Test {
    kind: Default,
    mode: Error,
    ident: "test_error",
    statement: Single(
        "SELECT",
    ),
    values: [
        [],
    ],
}
